cmd/controller/kube/client.go

```diff
@@ -272,59 +274,146 @@ func (c *Client) GetClusterInfo() (*ClusterInfo, error) {
 	var res ClusterInfo
 	// Try to find pods cidr from nodes.
 	for _, node := range c.index.nodesByName {
-		if node.Spec.PodCIDR != "" {
-			subnet, err := netip.ParsePrefix(node.Spec.PodCIDR)
-			if err != nil {
-				return nil, fmt.Errorf("parse node pod cidr: %w", err)
-			}
-			res.PodCidr = netip.PrefixFrom(subnet.Addr(), 16).String()
+		res.PodCidr = getPodCidrFromNodeSpec(node)
+		if len(res.PodCidr) > 0 {
 			break
 		}
 	}
 
-	for _, info := range c.index.ipsDetails {
-		// Find service cidr from clusterIP services.
-		if svc := info.Service; svc != nil && svc.Spec.Type == corev1.ServiceTypeClusterIP {
-			if svc.Spec.ClusterIP != "" && svc.Spec.ClusterIP != corev1.ClusterIPNone {
-				addr, err := netip.ParseAddr(svc.Spec.ClusterIP)
-				if err != nil {
-					return nil, fmt.Errorf("parse service cluster ip: %w", err)
-				}
-				cidr, err := addr.Prefix(16)
-				if err != nil {
-					return nil, fmt.Errorf("get cluster ip prefix: %w", err)
-				}
-				res.ServiceCidr = cidr.String()
-				if res.PodCidr != "" {
-					break
-				}
-			}
-		}
-		// Find pod cidr from pod if not found from nodes.
-		if res.PodCidr == "" {
-			if pod := info.PodInfo; pod != nil && pod.Pod != nil && pod.Pod.Status.PodIP != "" {
-				addr, err := netip.ParseAddr(pod.Pod.Status.PodIP)
-				if err != nil {
-					return nil, fmt.Errorf("parse pod addr: %w", err)
-				}
-				cidr, err := addr.Prefix(16)
+	// Find pod cidr from pod if not found from nodes
+	if len(res.PodCidr) == 0 {
+		for _, info := range c.index.ipsDetails {
+			if pod := info.PodInfo; pod != nil && pod.Pod != nil {
+				podCidr, err := getPodCidrFromPodSpec(pod.Pod)
 				if err != nil {
-					return nil, fmt.Errorf("get pod ip prefix: %w", err)
+					return nil, err
 				}
-				res.PodCidr = cidr.String()
-				if res.ServiceCidr != "" {
+				if len(res.PodCidr) > 0 {
+					res.PodCidr = podCidr
 					break
 				}
 			}
 		}
 	}
-	if res.PodCidr == "" && res.ServiceCidr == "" {
+
+	// Try to find service cidr from Kubernetes API output
+	serviceCidr, err := getServiceCidr(ctx, c.client, c.kvisorNamespace)
+	if err != nil {
+		return nil, err
+	}
+	res.ServiceCidr = serviceCidr
+
+	if len(res.PodCidr) == 0 && len(res.ServiceCidr) == 0 {
```

Comment:
I have re-worked this part because in dual-stack clusters we can have services with IPv4 and IPv6 so we cannot rely in a single service to derive the CIDR from the IP, we would need to iterate trough all the services which is a expensive operation.

The new way relies on Kubernetes API response when an invalid IP is created, for example: `The range of valid IPs is 10.45.0.0/16"`. We can retrieve the CIDR from this output message. Kubernetes does not expose the service CIDR in any API so this is the approach other projects are following like https://github.com/submariner-io/submariner-operator/blob/devel/pkg/discovery/network/generic.go#L106-L143